Sources/NIOExtrasPerformanceTester/HTTP1PerformanceTestFramework.swift

```diff
@@ -0,0 +1,204 @@
+
+    func setUp() throws {
+        self.group = MultiThreadedEventLoopGroup(numberOfThreads: System.coreCount)
+        self.serverChannel = try ServerBootstrap(group: self.group)
+            .serverChannelOption(ChannelOptions.socketOption(.so_reuseaddr), value: 1)
+            .childChannelInitializer { channel in
+                channel.pipeline.configureHTTPServerPipeline(withPipeliningAssistance: true).flatMap {
+                    channel.pipeline.addHandler(SimpleHTTPServer())
+                }
+            }.bind(host: "127.0.0.1", port: 0).wait()
+    }
+
+    func tearDown() {
+        try! self.serverChannel.close().wait()
+        try! self.group.syncShutdownGracefully()
+    }
+
+    func run() throws -> Int {
+        var reqs: [Int] = []
+        reqs.reserveCapacity(self.numberOfRepeats)
+        for _ in 0..<self.numberOfRepeats {
+            var requestHandlers: [RepeatedRequests] = []
+            requestHandlers.reserveCapacity(self.numberOfClients)
+            var clientChannels: [Channel] = []
+            clientChannels.reserveCapacity(self.numberOfClients)
+            for _ in 0 ..< self.numberOfClients {
+                let clientChannel = try! ClientBootstrap(group: self.group)
+                    .channelInitializer { channel in
+                        channel.pipeline.addHTTPClientHandlers().flatMap {
+                            let repeatedRequestsHandler = RepeatedRequests(numberOfRequests: self.requestsPerClient,
+                                                                           eventLoop: channel.eventLoop,
+                                                                           head: self.head)
+                            requestHandlers.append(repeatedRequestsHandler)
+                            return channel.pipeline.addHandler(repeatedRequestsHandler)
+                        }.flatMap {
+                            self.extraInitialiser(channel)
+                        }
+                    }
+                    .connect(to: self.serverChannel.localAddress!)
+                    .wait()
+                clientChannels.append(clientChannel)
+            }
+
+            var writeFutures: [EventLoopFuture<Void>] = []
+            for clientChannel in clientChannels {
+                clientChannel.write(NIOAny(HTTPClientRequestPart.head(self.head)), promise: nil)
+                writeFutures.append(clientChannel.writeAndFlush(NIOAny(HTTPClientRequestPart.end(nil))))
+            }
+            let allWrites = EventLoopFuture<Void>.andAllComplete(writeFutures, on: writeFutures.first!.eventLoop)
+            try! allWrites.wait()
+
+            let streamCompletedFutures = requestHandlers.map { rh in rh.completedFuture }
+            let requestsServed = EventLoopFuture<Int>.reduce(0, streamCompletedFutures, on: streamCompletedFutures.first!.eventLoop, +)
+            reqs.append(try! requestsServed.wait())
+        }
+        return reqs.reduce(0, +) / self.numberOfRepeats
+    }
```

Comment:
Ok, this test is doing a _lot_, and very little of it is what we want to measure.

We should investigate rewriting this benchmark in terms of `EmbeddedChannel`. That'll remove all the network overhead, all the threading context switching mess, and many of the future allocations.

Is there any reason not to make that change?

Reply:
I'll try doing this as an additional rather than replacement set of tests as you rightly point out they have value but test very different things.